Applications/Spire/Source/KeyBindings/TagNameBox.cpp

```diff
@@ -0,0 +1,196 @@
+            operation) {
+          m_transaction.push(RemoveOperation(operation.m_index,
+            operation.get_value()->get_name()));
+        },
+        [&] (const ListModel<std::shared_ptr<AdditionalTag>>::UpdateOperation&
+            operation) {
+          m_transaction.push(UpdateOperation(operation.m_index,
+            operation.get_previous()->get_name(),
+            operation.get_value()->get_name()));
+        });
+    }
+  };
+
+  struct AdditionalTagNameValueModel : ValueModel<QString> {
+    mutable UpdateSignal m_update_signal;
+    std::shared_ptr<ListModel<std::shared_ptr<AdditionalTag>>> m_list;
+    std::shared_ptr<ValueModel<std::shared_ptr<AdditionalTag>>> m_source;
+    scoped_connection m_connection;
+
+    AdditionalTagNameValueModel(
+      std::shared_ptr<ListModel<std::shared_ptr<AdditionalTag>>> list,
+      std::shared_ptr<ValueModel<std::shared_ptr<AdditionalTag>>> source)
+      : m_list(std::move(list)),
+        m_source(std::move(source)),
+        m_connection(m_source->connect_update_signal(
+          std::bind_front(&AdditionalTagNameValueModel::on_update, this))) {}
+
+    QValidator::State get_state() const override {
+      return m_source->get_state();
+    }
+
+    const QString& get() const override {
+      return m_source->get()->get_name();
+    }
+
+    QValidator::State test(const QString& value) const override {
+      if(auto i = find_name(value); i != m_list->end()) {
+        return QValidator::Acceptable;
+      }
+      return QValidator::Invalid;
+    }
+
+    QValidator::State set(const QString& value) override {
+      if(auto i = find_name(value); i != m_list->end()) {
+        return m_source->set(*i);
+      }
+      return QValidator::Invalid;
+    }
+
+    connection connect_update_signal(
+        const UpdateSignal::slot_type& slot) const override {
+      return m_update_signal.connect(slot);
+    }
+
+    auto find_name(const QString& name) const {
+      return std::find_if(m_list->begin(), m_list->end(),
+        [&] (const std::shared_ptr<AdditionalTag>& tag) {
+          return name == tag->get_name();
+        });
+    }
+
+    void on_update(const std::shared_ptr<AdditionalTag>& tag) {
+      m_update_signal(tag->get_name());
+    }
+  };
+}
+
+TagNameBox::TagNameBox(
+  std::shared_ptr<ListModel<std::shared_ptr<AdditionalTag>>> tags,
+  QWidget* parent)
+  : TagNameBox(std::move(tags),
+      std::make_shared<LocalValueModel<std::shared_ptr<AdditionalTag>>>(
+        std::make_shared<OrderFieldAdditionalTag>(Nexus::Tag(),
+          OrderFieldInfoTip::Model())), parent) {}
+
+TagNameBox::TagNameBox(
+    std::shared_ptr<ListModel<std::shared_ptr<AdditionalTag>>> tags,
+    std::shared_ptr<ValueModel<std::shared_ptr<AdditionalTag>>> current,
+    QWidget* parent)
+    : QWidget(parent),
+      m_tags(std::move(tags)),
+      m_current(std::move(current)) {
+  auto name_list_model = std::make_shared<AdditionalTagNameListModel>(m_tags);
```

Comment:
Use ```EnumBox``` instead of reimplementing it.

Reply:
`EnumBox<std::shared_ptr<AdditionalTag>>` can’t be instantiated because `EnumBox<T>::default_view_builder` uses `to_text`, but `to_text` lacks the version of `std::shared_ptr<AdditionalTag>`.

Reply:
```to_text``` is an overloaded function that uses static polymorphism to determine what function to call. There is no reason why it can't be used with an ```EnumBox```.